src/benchmarks/fragmentation.cpp

```diff
@@ -0,0 +1,985 @@
+ */
+/*
+ * fragmentation.cpp -- fragmentation benchmarks definitions
+ */
+
+#include <cassert>
+#include <cerrno>
+#include <cstring>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include "benchmark.hpp"
+#include "benchmark_worker.hpp"
+#include "file.h"
+#include "libpmemobj.h"
+#include "os.h"
+#include "poolset_util.hpp"
+
+ /*
+ * The factor used for PMEM pool size calculation, accounts for metadata,
+ * fragmentation and etc.
+ */
+#define FACTOR 2
+
+ /* The minimum allocation size that pmalloc can perform */
+#define ALLOC_MIN_SIZE 64
+
+ /* OOB and allocation header size */
+#define OOB_HEADER_SIZE 64
+
+/*
+ * mem_usage -- specifies how memory will grow
+ */
+enum mem_usage_type {
+	MEM_USAGE_UNKNOWN,
+	MEM_USAGE_FLAT,
+	MEM_USAGE_PEAK,
+	MEM_USAGE_RAMP
+};
+
+static float fragmentation = 0;
+static float fragmentation2 = 0;
+static float fragmentation3 = 0;
+
+/*
+ * prog_args -- benchmark specific arguments
+ */
+struct prog_args {
+	char *background_mem_usage_type_str; /* mem_usage_type: flat, peak, ramp */
+	char *scenario; /*test scenario name*/
+	size_t start_obj_size; /*initial object size*/
+	size_t max_obj_size; /*maximal object size*/
+	size_t growth_factor; /*parametr how fast object will grow in ramp mode*/
+	size_t peak_multiplier; /*multipier of poolsize for peak allocations*/
+	unsigned growth_interval; /*time after objects will grow in ramp mode*/
+	unsigned operation_time; /*maximal operation time*/
+	unsigned peak_lifetime; /*lifetime of peak allocations*/
+	unsigned peak_allocs; /*number of peak allocations*/
+	unsigned seed; /* seed for randomization */
+	bool rand; /* use random numbers */
+};
+
+struct frag_obj
+{
+	size_t block_size; /*size of pmemobj*/
+	PMEMoid oid;
+	bool is_allocated;
+	int op_index;
+
+	void init(struct prog_args *args)
+	{
+		block_size = args->start_obj_size;
+		is_allocated = false;
+		op_index = -1;
+	}
+};
+
+/*
+* frag_bench -- fragmentation benchmark context
+*/
+struct frag_bench
+{
+	PMEMobjpool *pop; /* persistent pool handle */
+	struct prog_args *pa; /* prog_args structure */
+	mem_usage_type background_mem_usage;
+	uint64_t n_ops;
+	size_t poolsize;
+	size_t theoretical_memory_usage;
+	struct frag_obj **pmemobj_array; /*array of objects used in benchmark*/
+	int(*func_op)(frag_obj * op_obj, frag_bench * fb,
+		struct frag_worker * fworker);
+
+	/*
+	* free all allocated pmemobj objects used in benchmark
+	*/
+	void free_pmemobj_array(unsigned nb)
+	{
+		for (unsigned i = 0; i < nb; i++)
+		{
+			if (pmemobj_array[i]->is_allocated)
+			{
+				pmemobj_free(&pmemobj_array[i]->oid);
+			}
+			free(pmemobj_array[i]);
+		}
+	}
+};
+
+/*
+* struct action_obj -- fragmentation benchmark action context
+*/
+struct action_obj {
+	PMEMoid *peak_oids = nullptr; /*oids array for peak allocations*/
+	bool peak_allocated = false;
+	unsigned allocation_start_time = 0; /*time of first allocation*/
+	unsigned deallocation_time = 0; /*time after object should be deallocated*/
```

Comment:
Is this tracked per-object or globally? From the code I think this is just a global timer on when to deallocate next object.
To simulate a real application, we must track object lifetime individually. My suggestion is keeping those lifetimes in a sorted collection, and then removing from the top of that collection.